frame/dex-router/src/benchmarking.rs

```diff
@@ -0,0 +1,135 @@
+use super::*;
+use crate::{self as pallet_dex_router, Pallet as DexRouter};
+use composable_traits::{
+	defi::CurrencyPair,
+	dex::{Amm, DexRouteNode},
+};
+use frame_benchmarking::{account, benchmarks, impl_benchmark_test_suite, whitelisted_caller};
+use frame_support::{
+	assert_ok,
+	traits::{fungibles::Mutate, OriginTrait},
+};
+use frame_system::{Origin, RawOrigin};
+use sp_arithmetic::Permill;
+
+benchmarks! {
+  where_clause { where
+	  T: pallet_dex_router::Config
+		 + pallet_curve_amm::Config
+		 + pallet_uniswap_v2::Config,
+	  <T as pallet_curve_amm::Config>::Balance: From<u128>, <T as pallet_curve_amm::Config>::AssetId: From<u128>,
+	  <T as pallet_uniswap_v2::Config>::Balance: From<u128>, <T as pallet_uniswap_v2::Config>::AssetId: From<u128>,
+	  <T as pallet_dex_router::Config>::Balance: From<u128>, <T as pallet_dex_router::Config>::AssetId: From<u128>, <T as pallet_dex_router::Config>::PoolId: From<u32>,
+  }
+  // benchmark inserting new route
+  update_route {
+	let owner = whitelisted_caller();
+	let amp_coeff = 5_u16;
+	let fee = Permill::zero();
+	let admin_fee = Permill::zero();
+	let usdt: <T as pallet_curve_amm::Config>::AssetId = 100_u128.into();
+	let usdc: <T as pallet_curve_amm::Config>::AssetId = 101_u128.into();
+	let assets = CurrencyPair::new(usdt, usdc);
+	let p = pallet_curve_amm::Pallet::<T>::do_create_pool(
+			&owner,
+			assets,
+			amp_coeff,
+			fee,
+			admin_fee
+		);
+	assert_ok!(&p);
+	// let stable_swap_pool_id : u128 = p.unwrap().into();
+	let eth: <T as pallet_uniswap_v2::Config>::AssetId = 102_u128.into();
+	let usdc: <T as pallet_uniswap_v2::Config>::AssetId = 101_u128.into();
+	let assets = CurrencyPair::new(usdc, eth);
+	let p = pallet_uniswap_v2::Pallet::<T>::do_create_pool(
+		&owner,
+		assets,
+		fee,
+		admin_fee
+		);
+	assert_ok!(&p);
+	// let uniswap_pool_id = p.unwrap();
+	let dex_route = vec![
+			DexRouteNode::Uniswap(0_u32.into()),
+			DexRouteNode::Curve(0_u32.into()),
+	];
+	let eth: <T as pallet_dex_router::Config>::AssetId = 102_u128.into();
+	let usdt: <T as pallet_dex_router::Config>::AssetId = 100_u128.into();
+	let currency_pair = CurrencyPair::new(eth, usdt);
+  } : _(RawOrigin::Signed(owner), currency_pair, Some(dex_route.clone().try_into().unwrap()))
+
+  exchange {
```

Comment:
Just for my understanding, is there a reason why this test does not have any execution range parameters eg: like here https://github.com/paritytech/substrate/blob/master/frame/examples/basic/src/benchmarking.rs#L40?

Reply:
I am not sure if that is required in general for benchmarks.
```
 14 // Executed Command:                                                                                                                                                       │  process didn't exit successfully: `/home/vivek/dev/        │
 13 // ./target/release/composable                                                                                                                                             │composable/target/debug/build/dali-runtime-acfdbd85040f8d9e/ │
 12 // benchmark                                                                                                                                                               │build-script-build` (exit status: 1)                         │
 11 // --chain=dali-dev                                                                                                                                                        │  --- stdout                                                 │
 10 // --execution=wasm                                                                                                                                                        │  ^[[1;33mInformation that should be included in a bug   @@@ │
  9 // --wasm-execution=compiled                                                                                                                                               └─────────────────────────────────────────────────────────────┘
  8 // --pallet=*                                                                                                                                                                                                                             
  7 // --extrinsic=*                                                                                                                                                           ┌─────────────────────────────────────────────────────────────┐
  6 // --steps=50                                                                                                                                                              │rust-analyzer failed to run build scripts:                   │
  5 // --repeat=20                                                                                                                                                             │error: failed to run custom build command for `dali-runtime  │
  4 // --raw                                                                                                                                                                   │v0.1.0 (/home/vivek/dev/composable/runtime/dali)`            │
  3 // --output=runtime/dali/src/weights                                                                                                                                       │                                                             │
  2 // --log                                                                                                                                                                   │Caused by:                                                   │
  1 // error  
  ```
In above command we specify how many times we want to repeat , and that should be enough right?

Reply:
According to my understanding the repeat serves a different purpose than what the execution range does. Execution range allows varying of complexity in designated steps for the extrinsic to estimate the correct weights as the range value increases. For example if there is complexity arising from the storage increasing with the number of calls to the function(even if the call is not taking the execution range as input), then the benchmark needs to simulate that as the goal is to estimate the weights taking into account the worst case. There are some examples of this in democracy pallet (https://github.com/ComposableFi/composable/blob/main/frame/democracy/src/benchmarking.rs#L177).

Reply:
I see, thanks for explaining this. IIUC this depends on values of parameters for extrinsic being tested for benchmark.
For dex-router benchmarks IMHO it does not matter much because
1) update_route complex case is to create new route so that is being tested for other operations like delete_route and update details of existing route complexity can be not be worst than inserting the new route.
2) for exchange and buy complexity should not change based on what amount is being exchanged.

Reply:
I see, thanks for the context
For 1. It seems the complexity also depends on the parameter passed in as the `dex_route` vector? i.e when it increases the function would take more time to execute(in reference to for loop in `do_update_route`). May be it does not affect much as the `T::MaxHopsInRoute` can't be that large anyway 
For 2. The complexity seems to depend on the number of `DexRouteNode`s returned by `get_route` for the given asset pair, which again seems to depend on `T::MaxHopsInRoute`.  But as you said it does not depend on the amount exchanged.